### DISCUSSION TITLE
ShouldBeUnique contract in PendingDispatch

### DISCUSSION THREAD

**Discussion by DimaVIII** [laravel/framework#48107] [Q&A]
Hello,

I'm trying to solve a issue for a package https://github.com/mateusjunges/trackable-jobs-for-laravel/issues/45. 

### Description
`__construct` function of a Job is triggered as a Job is processed by `Illuminate\Foundation\Bus\PendingDispatch`.
In this package a model is created in the constructor in order to inject the job tracking functionality: 
https://github.com/mateusjunges/trackable-jobs-for-laravel/blob/65ecdf71018982b79abfe5428e2679d5b96374d2/src/Concerns/Trackable.php#L20

### Issue
If a job implements the `ShouldBeUnique` interface all duplicated jobs are discarded by `PendingDispatch` but the constructor of the job is getting triggered in the process and a model is created even the job was discarded and never entered the queue.

### Question
I'm trying to find a way to delete the model from the database or prevent the creation of the model.

- There is no event fired, if the the job is discarded by `PendingDispatch`. https://github.com/laravel/framework/blob/81ae53fcdb1a552237ef73f65f1f3faac8236641/src/Illuminate/Foundation/Bus/PendingDispatch.php#L188
- There is no way to check the cache for a existing lock as this function runs later and locks can only be obtained not checked.
https://github.com/laravel/framework/blob/81ae53fcdb1a552237ef73f65f1f3faac8236641/src/Illuminate/Foundation/Bus/PendingDispatch.php#L163

**Comment by aprokopenko** [comment]
Just FYI, there is more issues with ShouldBeUnique, I'm working on some fixes for it: https://github.com/laravel/framework/issues/47761, https://github.com/laravel/framework/pull/47765. Need to fix unit tests after my updates

**Comment by henzeb** [comment]
You could place this logic in the destruct. That's how pending dispatch work.
